service-base/src/main/java/org/eclipse/hono/service/credentials/CredentialsHttpEndpoint.java

```diff
@@ -0,0 +1,135 @@
+     */
+    @Autowired
+    public CredentialsHttpEndpoint(final Vertx vertx) {
+        super(Objects.requireNonNull(vertx));
+    }
+
+    @Override
+    public void addRoutes(Router router) {
+
+        final String pathWithTenant = String.format("/%s/:%s", CredentialsConstants.CREDENTIALS_ENDPOINT, PARAM_TENANT_ID);
+        // ADD credentials
+        router.route(HttpMethod.POST, pathWithTenant).consumes(HttpEndpointUtils.CONTENT_TYPE_JSON)
+                .handler(this::doAddCredentialsJson);
+        router.route(HttpMethod.POST, pathWithTenant)
+                .handler(ctx -> HttpEndpointUtils.badRequest(ctx.response(), "missing or unsupported content-type"));
+    }
+
+    private static String getTenantParam(final RoutingContext ctx) {
+        return ctx.request().getParam(PARAM_TENANT_ID);
+    }
+
+    @Override
+    public String getName() {
+        return CredentialsConstants.CREDENTIALS_ENDPOINT;
+    }
+
+    private void doAddCredentialsJson(final RoutingContext ctx) {
+        try {
+            JsonObject payload = null;
+            if (ctx.getBody().length() > 0) {
+                payload = ctx.getBodyAsJson();
+            }
+            addCredentials(ctx, payload);
+        } catch (DecodeException e) {
+            HttpEndpointUtils.badRequest(ctx.response(), "body does not contain a valid JSON object");
+        }
+    }
+
+    private void addCredentials(final RoutingContext ctx, final JsonObject payload) {
+
+        if (payload == null) {
+            HttpEndpointUtils.badRequest(ctx.response(), "missing body");
+        } else {
+            Object deviceId = payload.getValue(FIELD_DEVICE_ID);
+            if (deviceId == null) {
+                HttpEndpointUtils.badRequest(ctx.response(), String.format("'%s' param is required", FIELD_DEVICE_ID));
+            } else if (!(deviceId instanceof String)) {
+                HttpEndpointUtils.badRequest(ctx.response(), String.format("'%s' must be a string", FIELD_DEVICE_ID));
+            } else {
+                final String tenantId = getTenantParam(ctx);
+                logger.debug("adding credentials for device [tenant: {}, device: {}, payload: {}]", tenantId, deviceId, payload);
+                final HttpServerResponse response = ctx.response();
+                final JsonObject requestMsg = CredentialsConstants.getServiceRequestAsJson(CredentialsConstants.OPERATION_ADD, tenantId, (String) deviceId, payload);
+                doAddCredentialsAction(ctx, requestMsg, (status, addCredentialsResult) -> {
+                    response.setStatusCode(status);
+                    switch(status) {
+                        case HttpURLConnection.HTTP_CREATED:
+                        response
+                                .putHeader(
+                                        HttpHeaders.LOCATION,
+                                        String.format("/%s/%s/%s", CredentialsConstants.CREDENTIALS_ENDPOINT, tenantId, deviceId));
```

Comment:
You haven't implemented the *get* method yet, but anyway this is probably not the right URI that a client can use to retrieve the resource that has been created by this operation. That is because a device can have arbitrary credentials registered for it.
The URI should probably look something
`/creadentials/$tenantId/$type/$authId`

Reply:
Yeah I know, the get will be in the next step :-) 
Maybe the address could be `/creadentials/$tenantId/$authId/$type`, following the same structure on how it's defined internally ?

Reply:
sure, be my guest :-)